pkg/terra/contract_config_tracker.go

```diff
@@ -3,41 +3,140 @@ package terra
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"strconv"
-
-	"github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
+	"sync"
+	"time"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/smartcontractkit/chainlink/core/utils"
 	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
 )
 
 var _ types.ContractConfigTracker = (*ContractTracker)(nil)
 
 type ContractTracker struct {
+	utils.StartStopOnce
 	jobID       string
 	address     sdk.AccAddress
 	chainReader client.Reader
 	log         Logger
+	cfg         Config
+	stop, done  chan struct{}
+
+	// cached state
+	mu             sync.RWMutex
+	ts             *time.Time
+	changedInBlock uint64
+	configDigest   types.ConfigDigest
+	contractConfig types.ContractConfig
 }
 
-func NewContractTracker(address sdk.AccAddress, jobID string, chainReader client.Reader, lggr Logger) *ContractTracker {
+func NewContractTracker(address sdk.AccAddress, jobID string, chainReader client.Reader, cfg Config, lggr Logger) *ContractTracker {
 	contract := ContractTracker{
 		jobID:       jobID,
 		address:     address,
 		chainReader: chainReader,
 		log:         lggr,
+		cfg:         cfg,
+		stop:        make(chan struct{}),
+		done:        make(chan struct{}),
 	}
 	return &contract
 }
 
+func (ct *ContractTracker) Start() error {
+	return ct.StartOnce("TerraContractTracker", func() error {
+		ct.log.Debugf("Starting")
+		go ct.pollState()
+		return nil
+	})
+}
+
+func (ct *ContractTracker) Close() error {
+	return ct.StopOnce("TerraContractTracker", func() error {
+		ct.log.Debugf("Stopping")
+		close(ct.stop)
+		<-ct.done
+		return nil
+	})
+}
+
+func (ct *ContractTracker) pollState() {
+	defer close(ct.done)
+	tick := time.After(0)
+	for {
+		select {
+		case <-ct.stop:
+			return
+		case <-tick:
+			ctx, cancel := utils.ContextFromChan(ct.stop)
+			done := make(chan struct{})
+			go func() {
```

Comment:
this goroutine is just to work around the fact that the ctx cancellation isn't respected right? I think we could still just do it in line since the latestConfigDetails call is still bounded by the http timeout. It would mean that shutting down could potentially take the length of the http timeout in the worse case, but that doesn't seem so bad actually and would ensure we never leak any goroutines and keep things simpler

Reply:
Yeah I don't like leaking goroutines, but I think we're stuck with that as the less bad choice, since it's critical to yield for shutdown to proceed properly. I do think it can be done more simply without this extra level though :+1: . Fortunately there aren't any external side effects, since they just update internal state.